Honor SOURCE_DATE_EPOCH in more cases

gioele opened the pull request:
This PR is an extension of the merged PR #3152, making sure that SOURCE_DATE_EPOCH is used everywhere the current time is serialized in a file.

@volth, @bmwiedemann: could you please test this to see if further reduces or eliminate problems with reproducible build on Linux?

@volth: The MacOS issues will need to be taken care of separately.

WRT Author names, `GetAuthor` already honors the user-configurable env variable `USER`.

Closes #3134
Fixes #2711

jtanx commented on contrib/fonttools/Makefile.am:
```diff
@@ -35,6 +35,8 @@ noinst_PROGRAMS = findtable pfadecrypt rmligamarks stripttc woff	\
 
 woff_LDADD = $(ZLIB_LIBS)
 dewoff_LDADD = $(ZLIB_LIBS)
+ttf2eps_LDADD = $(top_builddir)/Unicode/libgunicode.la \
```
Seems a bit heavy to add a dependency on these dlls for just the one function (and considering the standalone nature of the rest of the contrib tools)

gioele replied:
That dependency isn't already there only because there is a lot of code copied and pasted in `ttf2eps`. :) Same for `showttf`

We really should go though it and remove all the duplicated and outdated code. This would leave only the core ttf2eps logic there. That is the point of a `libfontforge`, isn't it? ;)

jtanx replied:
True, but I can also see the reasoning; they have no dependencies on anything else, so they can be shipped independently of FF itself. 

For example, if I'm reading this right, the [fontforge-extras](https://packages.debian.org/sid/fontforge-extras) package ships with no dependency on fontforge itself.

frank-trampe replied:
I forgot to comment here, but our repackaging dropped fontforge-extras. I'm hoping that Debian will follow suit.

jtanx commented on fontforge/macbinary.c:
```diff
@@ -1423,7 +1422,6 @@ static int DumpMacBinaryHeader(FILE *res,struct macbinaryheader *mb) {
 
 	/* Creation time, (seconds from 1/1/1904) */
     now = mactime();
-    time(&now);
```
What's up with this?

gioele replied:
Oops, I forgot to add the background story in the commit. :)

That `time(&now)` has been added by GW during a mass commit. If the comment is right then using `time` there is plainly wrong, as `time` counts from 1970. On the other hand, `mactime` counts from 1904, just like the comment says. In addition, in the following line the modification time is also stored and there only `mactime` is used.

I trust the comment and the overall context, thus removed that `time`. However I have put it in a separate comment, so that it will be easy to bisect should any problem arise.

jtanx commented on fontforge/print.c:
```diff
@@ -1155,23 +1156,11 @@ static void dump_pdfprologue(PI *pi) {
 	    tm->tm_year+1900, tm->tm_mon+1, tm->tm_mday, tm->tm_hour, tm->tm_min, tm->tm_sec );
-#ifdef _NO_TZSET
     fprintf( pi->out, "Z)\n" );
-#else
```
Why is this removed?

gioele replied:
That part of the code was there only to calculate and output the timezone offset. With no timezone there is no need to calculate it. ;)

BTW, I think it is broken on MacOS (can't remember the details, I had a brief look before realizing it could just be deleted).

frank-trampe commented:
Why is it necessary to drop timezone support?

bmwiedemann commented:
I think the plan is to always use UTC everywhere so that users with different timezone settings can still get the same result from fontforge calls.

gioele commented:
hi @volth, hi @bmwiedemann, I added some code to make sure that the env variable `USER` is preferred to GECOS fields if `SOURCE_DATE_EPOCH`. I cannot see any other externally controlled source of non-reproducibility. Does this work for you?

I did quite a bit of testing and everything seems to work fine when `SOURCE_DATE_EPOCH` is set as well as when it is not set.

Any objections to merging this?

volth commented:
I am not sure about Debian, but NixOS does not a set fake stable ```$USER``` for the build, and I anticipate the need to forge ```$USER``` may meet some objections in the community

gioele commented:
What does the build system of NixOS does when an author/user _must_ be set? Does it default to something?

Or is there an equivalent `SOURCE_DATE_EPOCH` for `USER`? Something like `SOURCE_USER`? If such a thing exist we can easily support it.

volth commented:
> What does the build system of NixOS does when an author/user must be set?

I guess it is something very new and fontforge-specific. Are the other build tools besides ```fontforge``` with the concept of "```author/user must be set```" ?

gioele commented:
> > What does the build system of NixOS does when an author/user must be set?
>
> I guess it is something very new and fontforge-specific. Are the other build tools besides fontforge with the concept of "author/user must be set" ?

Too many. :) tar, for example, must put something in the user/group of each file it records in a tar ball. Git/SVN/Hg/CVS also require you to have an user/email set and will store it in the metadata files.

I think that the more general problem (that surely has already been dealt with by NixOS) is: "How do you deal with env variables that end up being encoded in a compiled file?".

frank-trampe commented:
We can't drop timezone support without getting feedback from a lot of people. Let's either trigger UTC output when SOURCE_DATE_EPOCH is set or, if that doesn't work for some reason, set another environment variable like TIMESTAMP_TZ.

gioele commented:
hi @frank-trampe, you have a point: there is a slight chance of things breaking because of the missing timezone (I can picture a script that checks against the local time using strings instead of proper time routines). Assuming sloppy code out there seems a sensible default. ;)

I rebased the branch against master, reverted most of the commit that forced UTC and added some code to conditionally use UTC if `SOURCE_DATA_EPOCH` is set.

frank-trampe commented:
@gioele, I'd like to rebase (perhaps in a new branch and pull request) with the "Output UTC" and its revert dropped. I can do it, but it's easier for you to do since you own the branch.

gioele commented:
hi @frank-trampe, I rebased the branch and squashed away the "Output UTC" commit/revert commit.

h01ger commented:
the implemented logic and decissions look good to me - though I'm not familar with this code base :)

Thank you!

gioele commented:
hi @frank-trampe and @jtanx, are there any outstanding objections to merging this PR?

frank-trampe commented:
Go ahead!

frank-trampe commented:
Or I'll do it.

gioele commented:
Well, you've been faster. Thanks :)
